mlir/include/mlir/IR/CommonTypeConstraints.td

```diff
@@ -636,6 +636,10 @@ def AnyScalableVector : ScalableVectorOf<[AnyType]>;
 
 // Shaped types.
 
+class ShapedTypeOf<list<Type> allowedTypes> :
+  ShapedContainerType<allowedTypes, IsShapedTypePred, "shaped",
+                      "::mlir::ShapedType">;
```

Comment:
One problem with this is that memrefs could now be passed because they are shaped types. E.g.:
```mlir
// Should arith.addf allocate a new buffer?
%0 = arith.addf %a, %b : memref<?xf32>
```

I'm not quite sure how to phrase it, but I think we should allow only types that have "SSA" semantics. Not sure if "SSA" is the right word... Whether they are shaped types or not may not even be that important.

Reply:
Being a value type is not sufficient (e.g. we shouldn't allow adding tuples or structs). So scalar or shaped value type is probably the most useful/accurate definition.